File: sdk/js/src/crypto.ts
/**
 * Real crypto implementations for SPL token verification.
 * Uses only node:crypto (zero external deps).
 */

import { createHash, verify as cryptoVerify, hkdfSync, generateKeyPairSync, createPrivateKey, createPublicKey } from 'node:crypto';

function toHex(buf: Uint8Array): string {
  return Buffer.from(buf).toString('hex');
}

function concat(...bufs: Uint8Array[]): Uint8Array {
  return Buffer.concat(bufs);
}

/**
 * Verify an Ed25519 signature over a message.
 * @param message - The signed bytes (e.g. canonical SPL policy)
 * @param signatureHex - Hex-encoded Ed25519 signature (64 bytes)
 * @param publicKeyHex - Hex-encoded Ed25519 public key (32 bytes)
 * @returns true if signature is valid
 */
export function verifyEd25519(
  message: Uint8Array | string,
  signatureHex: string,
  publicKeyHex: string,
): boolean {
  try {
    const msgBuf = typeof message === 'string' ? Buffer.from(message, 'utf8') : Buffer.from(message);
    const sigBuf = Buffer.from(signatureHex, 'hex');
    const pubBuf = Buffer.from(publicKeyHex, 'hex');

    const keyObj = {
      key: Buffer.concat([
        // Ed25519 SPKI prefix (DER header for 32-byte public key)
        Buffer.from('302a300506032b6570032100', 'hex'),
        pubBuf,
      ]),
      format: 'der' as const,
      type: 'spki' as const,
    };

    return cryptoVerify(null, msgBuf, keyObj, sigBuf);
  } catch {
    return false;
  }
}

/**
 * SHA-256 hash of data.
 */
export function sha256(data: Uint8Array | string): Uint8Array {
  return new Uint8Array(createHash('sha256').update(data).digest());
}

/**
 * Verify a Merkle proof against a root hash.
 * @param leafData - The leaf data (will be SHA-256 hashed)
 * @param proof - Array of {hash: hex, position: 'left'|'right'}
 * @param rootHex - Expected root hash in hex
 * @returns true if proof is valid
 */
export function verifyMerkleProof(
  leafData: string,
  proof: Array<{ hash: string; position: 'left' | 'right' }>,
  rootHex: string,
): boolean {
  try {
    let current = sha256(leafData);

    for (const step of proof) {
      const sibling = new Uint8Array(Buffer.from(step.hash, 'hex'));
      if (step.position === 'right') {
        current = sha256(concat(current, sibling));
      } else {
        current = sha256(concat(sibling, current));
      }
    }

    return toHex(current) === rootHex;
  } catch {
    return false;
  }
}

/**
 * Hash a tuple (array of values) by JSON-serializing then SHA-256.
 */
export function hashTuple(tuple: any[]): string {
  const serialized = JSON.stringify(tuple);
  return toHex(sha256(serialized));
}

/**
 * Derive a service-specific Ed25519 keypair using HKDF-SHA256.
 * Provides unlinkability: different services see different public keys.
 *
 * @param masterKeyHex - 32-byte master private key seed (hex)
 * @param serviceDomain - Service identifier (e.g. "api.example.com")
 * @returns { publicKey: hex, privateKey: hex }
 */
export function deriveServiceKey(
  masterKeyHex: string,
  serviceDomain: string,
): { publicKey: string; privateKey: string } {
  const masterKey = Buffer.from(masterKeyHex, 'hex');
  const salt = Buffer.from('agent-safe-v1', 'utf8');
  const info = Buffer.from(serviceDomain, 'utf8');

  const derived = hkdfSync('sha256', masterKey, salt, info, 32);
  const seed = Buffer.from(derived);

  // Build Ed25519 private key from derived seed
  const pkcs8Prefix = Buffer.from('302e020100300506032b657004220420', 'hex');
  const privDer = Buffer.concat([pkcs8Prefix, seed]);
  const keyObj = createPrivateKey({ key: privDer, format: 'der', type: 'pkcs8' });
  const pubKey = createPublicKey(keyObj);
  const pubDer = pubKey.export({ type: 'spki', format: 'der' });
  const pubRaw = pubDer.subarray(12);

  return {
    publicKey: Buffer.from(pubRaw).toString('hex'),
    privateKey: seed.toString('hex'),
  };
}

/**
 * Verify a hash chain receipt.
 * Hashes preimage (chainLength - index) times and checks it equals commitment.
 */
export function verifyHashChain(
  commitment: string,
  preimageHex: string,
  index: number,
  chainLength: number,
): boolean {
  try {
    let current: Uint8Array = new Uint8Array(Buffer.from(preimageHex, 'hex'));
    const steps = chainLength - index;

    for (let i = 0; i < steps; i++) {
      current = sha256(current);
    }

    return toHex(current) === commitment;
  } catch {
    return false;
  }
}

File: sdk/rust/src/crypto.rs
use ed25519_dalek::{Signature, SigningKey, VerifyingKey};
use sha2::{Digest, Sha256};

/// Verify an Ed25519 signature over a message.
pub fn verify_ed25519(message: &[u8], signature_hex: &str, public_key_hex: &str) -> bool {
    let Ok(sig_bytes) = hex::decode(signature_hex) else { return false };
    let Ok(pub_bytes) = hex::decode(public_key_hex) else { return false };

    let Ok(sig) = Signature::from_slice(&sig_bytes) else { return false };
    let Ok(pub_arr): Result<[u8; 32], _> = pub_bytes.try_into() else { return false };
    let Ok(key) = VerifyingKey::from_bytes(&pub_arr) else { return false };

    key.verify_strict(message, &sig).is_ok()
}

/// SHA-256 hash of data.
pub fn sha256(data: &[u8]) -> Vec<u8> {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher.finalize().to_vec()
}

/// SHA-256 hash of data as hex string.
pub fn sha256_hex(data: &[u8]) -> String {
    hex::encode(sha256(data))
}

/// A step in a Merkle proof.
pub struct MerkleProofStep {
    pub hash: String,
    pub position: String, // "left" or "right"
}

/// Verify a Merkle proof for leaf_data against root_hex.
pub fn verify_merkle_proof(leaf_data: &str, proof: &[MerkleProofStep], root_hex: &str) -> bool {
    let mut current = sha256(leaf_data.as_bytes());

    for step in proof {
        let Ok(sibling) = hex::decode(&step.hash) else { return false };
        let mut hasher = Sha256::new();
        if step.position == "right" {
            hasher.update(&current);
            hasher.update(&sibling);
        } else {
            hasher.update(&sibling);
            hasher.update(&current);
        }
        current = hasher.finalize().to_vec();
    }

    hex::encode(&current) == root_hex
}

/// HMAC-SHA256 (used internally for HKDF).
fn hmac_sha256(key: &[u8], data: &[u8]) -> Vec<u8> {
    const BLOCK_SIZE: usize = 64;
    let mut key_block = [0u8; BLOCK_SIZE];
    if key.len() > BLOCK_SIZE {
        let h = sha256(key);
        key_block[..h.len()].copy_from_slice(&h);
    } else {
        key_block[..key.len()].copy_from_slice(key);
    }

    let mut ipad: Vec<u8> = (0..BLOCK_SIZE).map(|i| 0x36 ^ key_block[i]).collect();
    let mut opad: Vec<u8> = (0..BLOCK_SIZE).map(|i| 0x5c ^ key_block[i]).collect();

    ipad.extend_from_slice(data);
    let inner_hash = sha256(&ipad);
    opad.extend_from_slice(&inner_hash);
    sha256(&opad)
}

/// HKDF-SHA256 (RFC 5869) extract-and-expand. Zero external dependencies.
fn hkdf_sha256(ikm: &[u8], salt: &[u8], info: &[u8], length: usize) -> Vec<u8> {
    // Extract
    let salt = if salt.is_empty() { vec![0u8; 32] } else { salt.to_vec() };
    let prk = hmac_sha256(&salt, ikm);

    // Expand
    let mut out = Vec::new();
    let mut prev = Vec::new();
    let mut i = 1u8;
    while out.len() < length {
        let mut input = prev.clone();
        input.extend_from_slice(info);
        input.push(i);
        prev = hmac_sha256(&prk, &input);
        out.extend_from_slice(&prev);
        i += 1;
    }
    out.truncate(length);
    out
}

/// Derive a service-specific Ed25519 keypair using HKDF-SHA256.
/// Provides unlinkability: different services see different public keys.
pub fn derive_service_key(master_key_hex: &str, service_domain: &str) -> Result<(String, String), crate::types::SplError> {
    let master_key = hex::decode(master_key_hex)
        .map_err(|e| crate::types::SplError(format!("invalid master key hex: {e}")))?;
    let salt = b"agent-safe-v1";
    let info = service_domain.as_bytes();
    let seed_bytes = hkdf_sha256(&master_key, salt, info, 32);
    let seed: [u8; 32] = seed_bytes.try_into()
        .map_err(|_| crate::types::SplError("HKDF output size mismatch".into()))?;

    let signing_key = SigningKey::from_bytes(&seed);
    let verifying_key = signing_key.verifying_key();
    Ok((
        hex::encode(verifying_key.as_bytes()),
        hex::encode(signing_key.as_bytes()),
    ))
}

/// Verify a hash chain receipt.
pub fn verify_hash_chain(
    commitment: &str,
    preimage_hex: &str,
    index: usize,
    chain_length: usize,
) -> bool {
    let Ok(mut current) = hex::decode(preimage_hex) else { return false };
    let steps = chain_length - index;

    for _ in 0..steps {
        current = sha256(&current);
    }

    hex::encode(&current) == commitment
}
